wasm/src/lib.rs

```diff
@@ -53,8 +53,111 @@ unsafe extern "C" fn _iroha_wasm_dealloc(offset: *mut u8, len: usize) {
     let _box = Box::from_raw(core::slice::from_raw_parts_mut(offset, len));
 }
 
+mod sealed {
+    pub trait Sealed {}
+}
+
+/// Marks a type whose values can be queried from the host environment and passed
+/// as arguments to `entrypoint` of a smart contract
+pub trait QueryHost: sealed::Sealed {
+    /// Return requested [`Self`] object from the host
+    fn query() -> Self;
+}
+
+impl sealed::Sealed for AccountId {}
+impl QueryHost for AccountId {
+    fn query() -> Self {
+        #[cfg(not(test))]
+        use host::query_authority as host_query_authority;
+        #[cfg(test)]
+        use tests::_iroha_wasm_query_authority_mock as host_query_authority;
+
+        // Safety: ownership of the returned result is transferred into `_decode_from_raw`
+        unsafe { decode_with_length_prefix_from_raw(host_query_authority()) }
+    }
+}
+
+impl sealed::Sealed for Event {}
+impl QueryHost for Event {
+    fn query() -> Self {
+        #[cfg(not(test))]
+        use host::query_triggering_event as host_query_triggering_event;
+        #[cfg(test)]
+        use tests::_iroha_wasm_query_triggering_event_mock as host_query_triggering_event;
+
+        // Safety: ownership of the result is transferred into `_decode_from_raw`
+        unsafe { decode_with_length_prefix_from_raw(host_query_triggering_event()) }
+    }
+}
+
+impl sealed::Sealed for NeedsPermissionBox {}
+impl QueryHost for NeedsPermissionBox {
```

Comment:
@Arjentix Does this make sense?

Reply:
Interesting moment... Well, let's keep it for now

Reply:
Shouldn't we implement this for every sub-type of `Event` as well?

Reply:
Probably